craft_application/grammar.py

```diff
@@ -92,17 +92,59 @@ def process_part(
                 f"Invalid grammar syntax while processing '{key}' in '{part_yaml_data}': {e}"
             ) from e
 
-        # special cases:
-        # - scalar values should return as a single object, not in a list.
-        # - dict values should return as a dict, not in a list.
-        if key not in _NON_SCALAR_VALUES or key in _DICT_ONLY_VALUES:
-            processed_grammar = processed_grammar[0] if processed_grammar else None  # type: ignore[assignment]
-
-        part_yaml_data[key] = processed_grammar
+        part_yaml_data[key] = post_process_grammar(
+            processor, key, processed_grammar, part_yaml_data
+        )
 
     return part_yaml_data
 
 
+def post_process_grammar(
+    processor: GrammarProcessor,
+    key: str,
+    processed_grammar: list[Any],
+    part_yaml_data: dict[str, Any],
+) -> list[Any] | dict[str, str] | None:
+    """Post-process primitives returns by the grammar processor.
+
+    Special cases:
+    - scalar values should return as a single object, not in a list.
+    - dict values should return as a dict, not in a list.
+
+    :returns: the post-processed primitives
+    """
+    if processor.variant == Variant.FOR_VARIANT:
+        if key in _DICT_ONLY_VALUES:
+            return merge_processed_dict(processed_grammar, part_yaml_data)
+        if key not in _NON_SCALAR_VALUES:
+            return processed_grammar[0] if processed_grammar else None
+    elif key not in _NON_SCALAR_VALUES or key in _DICT_ONLY_VALUES:
+        return processed_grammar[0] if processed_grammar else None
+    return processed_grammar
+
+
+def merge_processed_dict(
+    processed_grammar: list[Any],
+    part_yaml_data: dict[str, Any],
+) -> dict[str, str] | None:
+    """Merge list of dicts in a single dict.
+
+    :raises: CraftValidationError if duplicate keys are found.
+    :returns: Merged dict.
+    """
+    processed_grammar_dict: dict[str, str] = {}
+    for d in processed_grammar:
+        for k, v in d.items():
+            if k in processed_grammar_dict:
+                raise CraftValidationError(
+                    f"Duplicate keys in processed dict '{k}' in '{part_yaml_data}'"
+                )
+            processed_grammar_dict.update({k: v})  # type: ignore[assignment]
```

Comment:
It'd be nice if we could use set intersections to generate a full list of duplicate keys and then error with all the duplicates listed.